### PR TITLE
Fix issue where changes to `@export` fields did not trigger updates

### LINKED ISSUE #13335: Apollo Client 4: Query using @export directive does not properly react to cache changes
### Issue Description

I've noticed an issue after updating Apollo to v4. I have several queries that rely on client-side cache variables using the `@client` and `@export(as: ...)` directives, like this:

```gql
query QueryWithExport($userId: String!) {
  userId @client @export(as: "userId")

  user(id: $userId) {
    id
    address {
      id
      random
    }
  }
}
```

These queries are not properly updating when the cache is modified. For example, if I try deleting a field using `client.cache.modify(...)`, the queries do not attempt to re-fetch the data. If I change the queries so they use passed variables instead of `@export`, then they do work as expected.

This was working fine in Apollo Client v3, so this seems like a regression.

### Link to Reproduction

https://stackblitz.com/edit/apollo-export-cache-issue?file=src%2Fmain.jsx

### Reproduction Steps

1. Open the reproduction, wait for the query results
2. Observe that 2 network requests are made, one for each query (count is below the evict button)
3. Press "evict address"
4. Observe that the query on the right that uses `@export` does not load new data
5. Observe that the query on the left that doesn't use `@export` loads new data
6. Observe that network requests only increases by 1, where if both queries were updating it should increase by 2

### `@apollo/client` version

4.2.7

### PR DESCRIPTION
**Pull request by jerelmiller**
Fixes #13335

We had a regression in 4.x with regards to `@export` fields where queries would sometimes not respond to cache updates. This happened when a field changed an object that depended on the variables keyed by the `@export` values.

```gql
query ($userId: ID!) {
  userId @client @export(as: "userId")
  user(id: $userId) {
    id
    name
  }
}
```

In this case, `ObservableQuery` subscribed to the cache using the initial empty variables value, but never resubscribed after export variables were known. This meant changes to data inside that `user` object never notified the query of the change.

This fix now emits a `resolvedVariables` value along with the notification to `ObservableQuery` so that `ObservableQuery` can update its `variables` and resubscribe to the cache using the export variables. I added a whole bunch of other tests to catch other potential edge cases to make sure we didn't miss this for other situations. 

## Summary by CodeRabbit

* **Bug Fixes**
  * Fixed reactivity issues affecting queries that use `@export` variables.
  * Cache updates now correctly trigger refreshed results when exported-variable fields change or are evicted.
  * Improved behavior for refetching, partial data, cache-and-network requests, and network-only requests.
  * Improved handling of streamed responses and variable changes during in-flight requests.

* **Tests**
  * Added coverage for cache, network, refetch, partial-result, and streaming scenarios involving `@export` variables.

### CODE REVIEW

File: src/__tests__/local-state/export.ts
Comment (jerelmiller) [inline src/__tests__/local-state/export.ts:1]: Only one test really mattered (the one with the issue link above it), but I wanted to make sure we covered other potential edge cases, so I added several that I didn't see covered elsewhere.

File: src/core/ObservableQuery.ts [hunk 1]
```diff
@@ -1932,16 +1932,54 @@ Did you mean to call refetch(variables) instead of refetch({ variables })?`,
...
+    // isEqualQuery(notification, this) to fail for future notifications since
+    // the notification.variables holds the stale variables value. In this case,
+    // we want to allow the notification through only if the current variables
+    // value matches the resolved variables.
+    //
+    // Note: the check for matching resolved variables typically kicks in for
+    // multi-emission fetches (such as defer, or cache-and-network, etc).
+    if (!equal(resolvedVariables, this.variables)) {
+      if (!isEqualQuery(notification, this)) {
+        return;
```
Comment (DaleSeo) [inline src/core/ObservableQuery.ts:1976]: When `resolvedVariables` already match `this.variables`, we skip `isEqualQuery` entirely, including its query-document check. That means a still-running disposable operation for a previous query could publish to this observable after `reobserve({ query: ... })` switches it to a new query with the same variables?

**Review by DaleSeo — APPROVED** [review]
Looks good! Just a question.

File: src/core/ObservableQuery.ts (hunk 1 above)
Comment (jerelmiller) [reply src/core/ObservableQuery.ts:1976]: Argh, good catch. Let me see if I can reproduce because clearly our test suite doesn't guard against that if that were the case.

Comment (jerelmiller) [reply src/core/ObservableQuery.ts:1976]: Ok its a bit more complicated that that, but I can reproduce. We unsubscribe from old requests when a new one starts (which should cancel the previous request) so we don't get notifications from the old query. However, there is one case where we keep the notification pipeline alive, which is when results are delivered from the cache, so if I `reobserve` with a new query + switch to `cache-only` fetch policy while the first query is still in-flight, it is possible for notifications on the old query to be delivered. 

Added a test and fix for that in [e61d599](https://github.com/apollographql/apollo-client/pull/13367/commits/e61d599938e17eb3d8992f408cc01e8e56d49c68)
